src/quest/serializer.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any, Dict, Protocol, TypeVar, Callable, Tuple, TypedDict
+from .utils import get_obj_name, import_object
+import inspect
+
+T = TypeVar('T')
+
+
+class TypeFactory(Protocol[T]):
+    def __call__(self, *args: Any, **kwargs: Any) -> T:
+        ...
+
+
+class SerializedData(TypedDict):
+    _ms_factory: str
+    args: Any
+    kwargs: Any
+
+
+class StepSerializer(Protocol):
+    async def serialize(self, obj: Any) -> Any:
+        ...
+
+    async def deserialize(self, data: Any) -> Any:
+        ...
+
+
+class TypeSerializer(Protocol[T]):
+    async def serialize(self, obj: T) -> Tuple[Callable[..., T], Tuple[Any, ...], Dict[str, Any]]:
+        ...
+
+
+class NoopSerializer(StepSerializer):
+    async def serialize(self, obj: Any) -> Any:
+        # object already JSON-serializable
+        return obj
+
+    async def deserialize(self, data: Any) -> Any:
+        return data
+
+
+class MasterSerializer(StepSerializer):
+    def __init__(self, type_serializers: Dict[type, TypeSerializer[Any]]):
+        self._type_serializers = type_serializers
+
+    async def serialize(self, obj: Any) -> Any:
+        # Check if it is a known type - directly serializable to JSON
+        if isinstance(obj, (int, float, str, bool, type(None))):
+            return obj
+        elif isinstance(obj, dict):
+            return {await self.serialize(k): await self.serialize(v) for k, v in obj.items()}
+        elif isinstance(obj, list):
+            return [await self.serialize(v) for v in obj]
+        elif isinstance(obj, tuple):
+            return tuple([await self.serialize(v) for v in obj])
+
+        # Check if custom serializer is registered for the object's type
+        obj_type = type(obj)
+        serializer = self._type_serializers.get(obj_type)
+        if serializer:
+            factory, args, kwargs = await serializer.serialize(obj)
+            if inspect.isfunction(factory) and factory.__name__ == "<lambda>":
+                raise ValueError("Factory should not be a lambda function.")
+            factory_name = get_obj_name(factory)
+            return SerializedData(
+                _ms_factory=factory_name,
+                args=await self.serialize(args),
+                kwargs=await self.serialize(kwargs)
+            )
+
+        # Default
+        return await NoopSerializer().serialize(obj)
```

Comment:
I think here we can just return obj - no need to instantiate a NoopSerializer to do it.